### PR TITLE
Parse modem UUID from JWT token 

### PR DESCRIPTION
**Pull request by SeppoTakalo**
Default JWT tokens contain modem UUID in "iss" field.
For example:
{
  "iss": "nRF9160.50363154-3930-4d96-80f9-13121cf35adc",
  "jti": "..."
}
This is easiest way to fetch it.

But in order to issue default AT%JWT command, I needed some small fixes for the existing library as well:
JWT library should allow user to leave out sec_tag and expiration times.

AT%JWT and AT%JWT=0 and AT%JWT=0,0 and AT%JWT=,, AT%JWT=0,0,"" AT%JWT=0,0,"",""
returns only "iss" and "jti" fields and signed with device identity key

So numerical values can be zero or missing. Strings can be zero length or missing.

Sec_tag cannot be zero, it must be a valid tag, or missing in the parameter field.
When sec_tag is given, it key type must be given as well. When sec_tag is given as zero, I thread it as a missing field.

### CODE REVIEW

File: include/modem/modem_jwt.h [hunk 1]
```diff
@@ -76,6 +76,20 @@ struct jwt_data {
...
+ * This helper function utilizes JWT library to decode UUID from
+ * modem generated tokens. If succesfull a string is returned that
+ * contains UUID. Any error leads to this function returning NULL.
+ *
+ * Returned string consumes memory, which must later on be freed with
+ * @ref modem_jwt_free.
+ *
+ * @return string containing UUID or NULL on failure.
+ */
+const char *modem_jwt_get_uuid();
```
Comment (jayteemo) [inline include/modem/modem_jwt.h:91]: could you make the function signature like this one: https://github.com/nrfconnect/sdk-nrf/blob/8f5922cb5cd660612975d8226e5b52d3d318cce7/include/modem/modem_attest_token.h#L128

that way the user can get the device uuid and/or the fw uuid.  also the user won't need to free memory.
you can re-use the UUID structs/defines.

File: lib/modem_jwt/modem_jwt.c [hunk 2]
```diff
@@ -148,6 +158,43 @@ int modem_jwt_generate(struct jwt_data *const jwt)
...
+
+	int rc = modem_jwt_generate(&jwt);
+	if (rc) {
+		modem_jwt_free(jwt.jwt_buf);
+		return NULL;
+	}
+
+	char *p = strchr(jwt.jwt_buf, '.') + 1;
+	char *end = strrchr(jwt.jwt_buf, '.');
+	char *payload =  k_calloc(1, 256);
```
Comment (jayteemo) [inline lib/modem_jwt/modem_jwt.c:173]: you could use a macro such as:
https://github.com/nrfconnect/sdk-nrf/blob/8f5922cb5cd660612975d8226e5b52d3d318cce7/lib/modem_attest_token/modem_attest_token.c#L18

to determine the proper  base64 payload size

File: lib/modem_jwt/modem_jwt.c [hunk 3]
```diff
@@ -148,6 +158,43 @@ int modem_jwt_generate(struct jwt_data *const jwt)
...
+	size_t len;
+
+	rc = base64_decode(payload, 256, &len, p, end - p);
+	if (rc) {
+		k_free(payload);
+		modem_jwt_free(jwt.jwt_buf);
+		return NULL;
+	}
+	const char *iss = "iss\":\"nRF9160.";
```
Comment (jayteemo) [inline lib/modem_jwt/modem_jwt.c:182]: it would be more future-proof if you did not include the device name in this search.

**Comment by jayteemo** [comment]
> Sec_tag cannot be zero, it must be a valid tag, or missing in the parameter field.
> When sec_tag is given, it key type must be given as well. When sec_tag is given as zero, I thread it as a missing field.

it would be helpful to add some of this information as comments on either the `sec_tag` definition in the `jwt_data` or the `modem_jwt_generate` function.  right now, i'm not sure how useful it is to the user to sign something with the identity key, but they should be aware of what happens if they use a `sec_tag` of zero.

**Comment by SeppoTakalo** [comment]
@jayteemo Updated the PR as you suggested.

File: lib/modem_jwt/modem_jwt.c [hunk 4]
```diff
@@ -148,6 +159,76 @@ int modem_jwt_generate(struct jwt_data *const jwt)
 	return ret;
 }
 
+int copy_uuid_str(char *dst, const char *src, const char *field)
```
Comment (jayteemo) [inline lib/modem_jwt/modem_jwt.c:162]: add `static`

File: lib/modem_jwt/modem_jwt.c [hunk 5]
```diff
@@ -148,6 +159,76 @@ int modem_jwt_generate(struct jwt_data *const jwt)
...
+	char *dot = strchr(tag, '.');
+	if (!end || !dot || dot > end) {
+		return EINVAL;
+	}
+	char *uuid = dot + 1;
+	size_t len = end - uuid;
+	if (len < NRF_UUID_V4_STR_LEN) {
+		return EINVAL;
+	}
+	strncpy(dst, uuid, NRF_UUID_V4_STR_LEN);
```
Comment (jayteemo) [inline lib/modem_jwt/modem_jwt.c:184]: you could change this to: `strncpy(dst, uuid, NRF_UUID_V4_STR_LEN + 1);`
then delete the following line.

File: lib/modem_jwt/modem_jwt.c [hunk 6]
```diff
@@ -148,6 +159,76 @@ int modem_jwt_generate(struct jwt_data *const jwt)
...
+	start = strchr(jwt.jwt_buf, '.') + 1;
+	end = strrchr(jwt.jwt_buf, '.');
+	slen = end - start;
+	payload =  k_calloc(1, GET_BASE64_DECODED_LEN(slen));
+	if (!payload) {
+		rc = ENOMEM;
+		goto end;
+	}
+
+	rc = base64_decode(payload, 256, &len, start, slen);
```
Comment (jayteemo) [inline lib/modem_jwt/modem_jwt.c:211]: use actual buffer size.

File: lib/modem_jwt/modem_jwt.c [hunk 7]
```diff
@@ -148,6 +159,76 @@ int modem_jwt_generate(struct jwt_data *const jwt)
...
+
+	int rc = modem_jwt_generate(&jwt);
+	if (rc) {
+		goto end;
+	}
+
+	start = strchr(jwt.jwt_buf, '.') + 1;
+	end = strrchr(jwt.jwt_buf, '.');
+	slen = end - start;
+	payload =  k_calloc(1, GET_BASE64_DECODED_LEN(slen));
```
Comment (jayteemo) [inline lib/modem_jwt/modem_jwt.c:205]: i just noticed that i don't even use the macro i previously mentioned 🤦 
the "correct" way to determine the buffer size is to call `base64_decode` with  *dst = NULL or dlen = 0...
[for example](https://github.com/nrfconnect/sdk-nrf/blob/4544fb97667ea2b4cbc44ad4b8ad4123e8612132/lib/modem_attest_token/modem_attest_token.c#L247).
sorry about that.

File: lib/modem_jwt/modem_jwt.c (hunk 5 above)
Comment (SeppoTakalo) [reply lib/modem_jwt/modem_jwt.c:184]: Actually no.
`strncpy()` does not add NUL termination if the source is longer than `count` parameter.
https://en.cppreference.com/w/c/string/byte/strncpy

In this case, we are copying substring from a longer JSON string, so I need to terminate it.

`strlcpy()` would work like you suggest, but that is BSD extension. It is part of Newlib, but requires `__BSD_VISIBLE` flag, and I have not seen anyone using that, so I'll stay C99 compatible.

File: lib/modem_jwt/modem_jwt.c (hunk 6 above)
Comment (SeppoTakalo) [reply lib/modem_jwt/modem_jwt.c:211]: Thanks. Good catch.

File: lib/modem_jwt/modem_jwt.c (hunk 4 above)
Comment (SeppoTakalo) [reply lib/modem_jwt/modem_jwt.c]: Fixed.

**Comment by SeppoTakalo** [comment]
I feel like our compliance checks script is missconfigured.
It is complaining for me that my Git message body contains lines that are longer that 72 character.
Such rule does not exist in https://docs.zephyrproject.org/latest/contribute/#commit-guidelines
The limitation should only apply to subject line.

**Comment by rlubos** [comment]
> I feel like our compliance checks script is missconfigured.
> It is complaining for me that my Git message body contains lines that are longer that 72 character.
> Such rule does not exist in https://docs.zephyrproject.org/latest/contribute/#commit-guidelines
> The limitation should only apply to subject line.

Indeed it doesn't seem to be explicitly stated in the documentation, however this rule has been followed since the beginning. In NCS we just copy the gitlint configuration from Zephyr.

**Comment by rlubos** [comment]
@jayteemo Are there any integration tests that could run on this PR?

File: lib/modem_jwt/modem_jwt.c (hunk 5 above)
Comment (jayteemo) [reply lib/modem_jwt/modem_jwt.c:193]: ah yes, `uuid` isn’t null terminated here 👍

**Comment by jayteemo** [comment]
> @jayteemo Are there any integration tests that could run on this PR?

At the moment, I don't know if there is anything applicable.
This new UUID function will be used by the nrf_cloud lib when the config option `NRF_CLOUD_CLIENT_ID_SRC_INTERNAL_UUID` is enabled.

**Review by jayteemo — APPROVED** [review]

**Comment by SeppoTakalo** [comment]
I'll create some unittests for this, so we have at least some coverage. I already have at least one dummy token that I can use to verify.

**Review by rlubos — APPROVED** [review]

File: include/modem/modem_jwt.rst [hunk 8]
```diff
@@ -17,12 +17,12 @@ To use the library to request a JWT, complete the following steps:
 
 You can configure the following values in the :c:struct:`jwt_data` structure:
 
-* :c:member:`jwt_data.sec_tag` - Required. The sec tag must contain a valid signing key.
-* :c:member:`jwt_data.key` - Required. Defines the type of key in the sec tag.
-* :c:member:`jwt_data.alg` - Required. Defines the JWT signing algorithm. Currently, only ECDSA 256 is supported.
+* :c:member:`jwt_data.sec_tag` - Optional. The sec tag must contain a valid signing key. If zero is given, modem uses it's own key for signing.
```
Comment (umapraseeda) [inline include/modem/modem_jwt.rst:20]:
```suggestion
* :c:member:`jwt_data.sec_tag` - Optional. The sec tag must contain a valid signing key. If zero is given, modem uses its own key for signing.
```

**Comment by SeppoTakalo** [comment]
@jayteemo  and @rlubos I appended this PR with a commit that contains unittests for JWT library.

**Review by rlubos — APPROVED** [review]

**Review by umapraseeda — APPROVED** [review]

**Comment by SeppoTakalo** [comment]
How can I supply a JWT token in a unittest?
```
-:347: WARNING:MISSING_SPACE: break quoted strings at a space character
#347: FILE: tests/lib/modem_jwt/src/jwt_test.c:37:
+				"***"
+				"GJlNDg1ZmY4ZGIyMWY1MWRmZTcyNDJlN2NhN2I5NDU2YzkwZjY5ODE1ZTc3ZWE2MW"
```
Tokens don't have spaces, so it is a bit hard to obey that rule.

**Comment by rlubos** [comment]
> How can I supply a JWT token in a unittest?
> 
> ```
> -:347: WARNING:MISSING_SPACE: break quoted strings at a space character
> #347: FILE: tests/lib/modem_jwt/src/jwt_test.c:37:
> +				"***"
> +				"GJlNDg1ZmY4ZGIyMWY1MWRmZTcyNDJlN2NhN2I5NDU2YzkwZjY5ODE1ZTc3ZWE2MW"
> ```
> 
> Tokens don't have spaces, so it is a bit hard to obey that rule.

AFAIR checkpatch won't complain if such "unsplittable" string exceeds the char/line limit, so one option would be to put the entire token into a single line. Not the nicest solution, but I don't know if there's actually a way to instruct checkpatch to ignore that particular warning, perhaps someone else knows better.

**Comment by SeppoTakalo** [comment]
```
-:345: WARNING:LONG_LINE: line length of 517 exceeds 100 columns
#345: FILE: tests/lib/modem_jwt/src/jwt_test.c:35:
+static const char jwt_token[] = "%JWT: \"***"";
```

It does complain.. Now I don't know how to workaround it.

**Comment by rlubos** [comment]
> ```
> -:345: WARNING:LONG_LINE: line length of 517 exceeds 100 columns
> #345: FILE: tests/lib/modem_jwt/src/jwt_test.c:35:
> +static const char jwt_token[] = "%JWT: \"***"";
> ```
> 
> It does complain.. Now I don't know how to workaround it.

Ehh, one of the checkpatch oddiness... it seems that it doesn't treat string with internal quotes as unsplittable... The following configuration silenced it when I tested locally:
```
static const char jwt_token[] =
	"%JWT: \""
	"eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NiIsImtpZCI6ImU2YzM1Y2VhMWY1NDFhNGJlNDg1ZmY4ZGIyMWY1MWRmZTcyNDJlN2NhN2I5NDU2YzkwZjY5ODE1ZTc3ZWE2MWYifSAg.eyJpc3MiOiJuUkY5MTYwLjUwMzYzMTU0LTM5MzAtNGQ5Ni04MGY5LTEzMTIxY2YzNWFkYyIsImp0aSI6Im5SRjkxNjAuZjcyZDk3ODktMmU3My00NzgwLTkyNDgtYjM5NTJmODliODcyLjk0MmQ5OWI5MDNhY2EyYzNkZDY0OWI1ZDYxMjMyMmRjIiwic3ViIjoiNTAzNjMxNTQtMzkzMC00ZDk2LTgwZjktMTMxMjFjZjM1YWRjIn0g.tW2wG-sSE01jtrcT9JmFvPVIvEB99jUGGHIUCCH65fG5HtBrisOD1cl0L49cuyo2OE6XXvI9wdKSASxDQPJ8lw"
	"\"";
```

**Pull request merged**